src/wallet/scriptpubkeyman.cpp

```diff
@@ -1635,4 +1635,212 @@ bool DescriptorScriptPubKeyMan::CanUpdateToWalletDescriptor(const WalletDescript
+        return ISMINE_SPENDABLE;
+    }
+    return ISMINE_NO;
+}
+
+isminetype SilentPaymentDescriptorScriptPubKeyMan::IsMine(const std::vector<XOnlyPubKey>& output_keys, const bip352::PublicData& public_data)
+{
+    LOCK(cs_desc_man);
+    if (m_wallet_descriptor.descriptor->GetOutputType() != OutputType::SILENT_PAYMENTS) {
+        return ISMINE_NO;
+    }
+
+    const auto provider{GetSPProvider()};
+    auto tweaks{bip352::ScanForSilentPaymentOutputs(
+        provider.sp_keys.first, public_data, provider.sp_keys.second, output_keys, m_map_label_tweaks)};
+    if (!tweaks.has_value()) {
+        return ISMINE_NO;
+    }
+    WalletBatch batch(m_storage.GetDatabase());
+    if (!batch.TxnBegin()) {
+        throw std::runtime_error(strprintf("Error during descriptors tweak top up. Cannot start db transaction wallet %s", m_storage.GetDisplayName()));
+    }
+    for (const auto& tweak : *tweaks) {
+        if (!TopUpWithDB(batch, tweak.tweak)) {
+            throw std::runtime_error(std::string(__func__) + ": writing tweak failed");
+        }
+    }
+    if (!batch.TxnCommit()) {
+        throw std::runtime_error(strprintf("Error during descriptors tweak top up. Cannot commit changes for wallet %s", m_storage.GetDisplayName()));
+    }
+    return ISMINE_SPENDABLE;
+}
+
+util::Result<CTxDestination> SilentPaymentDescriptorScriptPubKeyMan::GetReservedDestination(const OutputType type, bool internal, int64_t& index)
+{
+    LOCK(cs_desc_man);
+    // Destination at index 0 is the change destination
+    auto op_dest = GetLabeledDestination(0);
+    return CTxDestination{op_dest};
+}
+
+bool SilentPaymentDescriptorScriptPubKeyMan::TopUp(const uint256& tweak)
+{
+    WalletBatch batch(m_storage.GetDatabase());
+    if (!batch.TxnBegin()) return false;
+    bool res = TopUpWithDB(batch, tweak);
+    if (!batch.TxnCommit()) throw std::runtime_error(strprintf("Error during descriptors tweak top up. Cannot commit changes for wallet %s", m_storage.GetDisplayName()));
+    return res;
+}
+
+bool SilentPaymentDescriptorScriptPubKeyMan::TopUpWithDB(WalletBatch& batch, const uint256& tweak)
+{
+    LOCK(cs_desc_man);
+    assert(m_wallet_descriptor.descriptor->GetOutputType() == OutputType::SILENT_PAYMENTS); // Should only ever be called on silent payment descriptorspkman
+
+    const auto provider{GetSPProvider()};
+    CPubKey tweaked_pub{provider.sp_keys.second};
+    tweaked_pub.TweakAdd(tweak.data());
+    const auto spk = GetScriptForDestination(WitnessV1Taproot{XOnlyPubKey{tweaked_pub}});
+    m_map_spk_tweaks.emplace(spk, tweak);
+    m_storage.TopUpCallback({spk}, this);
+    return batch.WriteSilentPaymentsTweak(GetID(), tweak);
+}
+
+FlatSigningProvider SilentPaymentDescriptorScriptPubKeyMan::GetSPProvider() const
+{
+    AssertLockHeld(cs_desc_man);
+    FlatSigningProvider provider;
+    provider.keys = GetKeys();
+    FlatSigningProvider tmp;
+    m_wallet_descriptor.descriptor->ExpandPrivate(0, provider, tmp);
+    return tmp;
+}
```

Comment:
Can likely be removed if we have custom methods for storing scan/spend pubkey data, or custom methods for getting them.